server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1595,16 +1600,26 @@ private void innerOpenEngineAndTranslog(LongSupplier globalCheckpointSupplier) t
         assert recoveryState.getRecoverySource().expectEmptyRetentionLeases() == false || getRetentionLeases().leases().isEmpty()
             : "expected empty set of retention leases with recovery source [" + recoveryState.getRecoverySource()
             + "] but got " + getRetentionLeases();
-        synchronized (mutex) {
-            verifyNotClosed();
-            assert currentEngineReference.get() == null : "engine is running";
+        synchronized (engineMutex) {
```

Comment:
a general idea here is to randomly add a warmer that actually does some work in our tests. I am a little concerned that we might change state or things like that concurrently that will cause issues. If we randomly do work in the opening phase that might surface these issues.

Reply:
I pushed https://github.com/elastic/elasticsearch/pull/45263/commits/681696020c666cdbf29d3743a38eaeb332c4993a to add an engine warmer to IndexShardTestCase. I also tried to add this warmer to integration tests but that would require production code changes. I can do it in a follow-up.

Reply:
👍